velox/functions/prestosql/Zip.cpp

```diff
@@ -0,0 +1,221 @@
+  VELOX_USER_CHECK_GT(
+      inputArgs.size(), 1, "zip requires at least two parameters");
+
+  for (auto& arg : inputArgs) {
+    VELOX_USER_CHECK_EQ(
+        arg->type()->kind(),
+        TypeKind::ARRAY,
+        "zip requires arguments of type ARRAY");
+  }
+}
+
+class ZipFunction : public exec::VectorFunction {
+  static const auto kMinArity = 2;
+  static const auto kMaxArity = 7;
+
+ public:
+  /// This class implements the zip function.
+  ///
+  /// DEFINITION:
+  /// zip(ARRAY[T], ARRAY[U]) -> ARRAY(ROW[T,U])
+  /// where we create a ROW[Ti, Ui] for every ith element in ARRAY[T], ARRAY[U].
+  /// The smaller array is padded with nulls.
+  ///
+  /// IMPLEMENTATION:
+  ///  1. The general idea is to create a new dictionary vector for each input
+  ///  array vector and enumerate their indices to create a 1:1 mapping.
+  ///  2. To do this, for each row we determine which is the largest Array
+  ///  and subsequently pad the smaller arrays with nulls.
+  ///  3. Then we take the resultant padded vectors together and create one ROW
+  ///  Vector.
+  ///  4. This forms the base to create the final output Array vector, whose
+  ///  Arrays are the size of the largest input Array.
+  ///
+  ///  Note:
+  ///   - We make no copy's of any constituent elements and are agnostic to
+  ///   types.
+  ///   - For compatibility with Presto a maximum arity of 7 is enforced.
+
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      const TypePtr& outputType,
+      exec::EvalCtx* context,
+      VectorPtr* result) const override {
+    validateInputTypes(args);
+    const vector_size_t numInputArrays = args.size();
+
+    exec::DecodedArgs decodedArgs(rows, args, context);
+    std::vector<const ArrayVector*> baseVectors(numInputArrays);
+    std::vector<const vector_size_t*> rawSizes(numInputArrays);
+    std::vector<const vector_size_t*> rawOffsets(numInputArrays);
+    std::vector<const vector_size_t*> indices(numInputArrays);
+
+    for (int i = 0; i < numInputArrays; i++) {
+      baseVectors[i] = decodedArgs.at(i)->base()->as<ArrayVector>();
+      rawSizes[i] = baseVectors[i]->rawSizes();
+      rawOffsets[i] = baseVectors[i]->rawOffsets();
+      indices[i] = decodedArgs.at(i)->indices();
+    }
+
+    // Size of elements in result vector.
+    vector_size_t resultElementsSize = 0;
+    auto pool = context->pool();
+
+    // Determine what the size of the resultant elements will be so we can
+    // reserve enough space.
+    auto getMaxArraySize = [&](vector_size_t row) -> vector_size_t {
+      vector_size_t maxSize = 0;
+      for (int i = 0; i < numInputArrays; i++) {
+        maxSize = std::max(maxSize, rawSizes[i][indices[i][row]]);
+      }
+      return maxSize;
+    };
+
+    BufferPtr maxArraySizesBuffer = allocateSizes(rows.end(), pool);
+    auto maxArraySizes = maxArraySizesBuffer->asMutable<vector_size_t>();
+    rows.applyToSelected([&](auto row) {
```

Comment:
This loop might be more efficient if vectorized, e.g. initialize sizes buffer by copying sizes buffer from the first input array, then loop over remaining input arrays. For each input array loop over rows and update sizes[row] if it is less than current array's size. At the end run another loop over rows to compute the resultElementsSize by adding up all sizes.